crates/ra_ide_api/src/symbol_index.rs

```diff
@@ -137,13 +136,32 @@ impl Hash for SymbolIndex {
 
 impl SymbolIndex {
     fn new(mut symbols: Vec<FileSymbol>) -> SymbolIndex {
-        fn cmp(s1: &FileSymbol, s2: &FileSymbol) -> Ordering {
-            unicase::Ascii::new(s1.name.as_str()).cmp(&unicase::Ascii::new(s2.name.as_str()))
+        fn cmp_key<'a>(s1: &'a FileSymbol) -> impl Ord + Eq + 'a {
+            unicase::Ascii::new(s1.name.as_str())
         }
-        symbols.par_sort_by(cmp);
-        symbols.dedup_by(|s1, s2| cmp(s1, s2) == Ordering::Equal);
-        let names = symbols.iter().map(|it| it.name.as_str().to_lowercase());
-        let map = fst::Map::from_iter(names.zip(0u64..)).unwrap();
+
+        symbols.par_sort_by(|s1, s2| cmp_key(s1).cmp(&cmp_key(s2)));
+
+        let mut builder = fst::MapBuilder::memory();
+
+        let mut last_batch_start = 0;
+
+        for idx in 0..symbols.len() {
+            if symbols.get(last_batch_start).map(cmp_key) == symbols.get(idx + 1).map(cmp_key) {
+                continue;
```

Comment:
Note that we loose the last batch here. Grouping with a loop is tricky, but there's group_by in itertools: https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.group_by

Reply:
Sorry, I don't quite get what you mean by losing the last batch.

Re: itertools: nice, it didn't occur to me to check if we were using it.

I'm a bit unsure on how to make it work, since `group_by` has a HRTB in the `key` callback, and our `cmp_key` is borrowed... I found a similar problem trying to call `symbols.par_sort_by_key` instead of `symbols.par_sort_by`.

Reply:
Ah, we don’t loose the last batch because we use idx + 1, missed that on the first reading, cool!

And yeah, I don’t know of a group by equivalent that works with borrowed keys...